gpcontrib/gp_orca/gporca.c

```diff
@@ -1,14 +1,86 @@
 #include "postgres.h"
 
 #include "cdb/cdbvars.h"
+#include "executor/executor.h"
 #include "miscadmin.h"
+#include "optimizer/orca.h"
+#include "optimizer/planner.h"
 #include "utils/builtins.h"
+#include "utils/guc.h"
+#include "utils/memutils.h"
 
 PG_MODULE_MAGIC;
 
+extern void InitGPOPT();
+extern void TerminateGPOPT();
+
+extern void compute_jit_flags(PlannedStmt* pstmt);
+
 void		_PG_init(void);
 void		_PG_fini(void);
 
+static planner_hook_type prev_planner = NULL;
+
+static PlannedStmt *
+gp_orca_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
+{
+	PlannedStmt	   *result = NULL;
+
+	/*
+	 * Use ORCA only if it is enabled and we are in a coordinator QD process.
+	 *
+	 * ORCA excels in complex queries, most of which will access distributed
+	 * tables. We can't run such queries from the segments slices anyway because
+	 * they require dispatching a query within another - which is not allowed in
+	 * GPDB (see querytree_safe_for_qe()). Note that this restriction also
+	 * applies to non-QD coordinator slices.  Furthermore, ORCA doesn't currently
+	 * support pl/<lang> statements (relevant when they are planned on the segments).
+	 * For these reasons, restrict to using ORCA on the coordinator QD processes only.
+	 *
+	 * PARALLEL RETRIEVE CURSOR is not supported by ORCA yet.
+	 */
+	if (optimizer &&
+		GP_ROLE_DISPATCH == Gp_role &&
+		IS_QUERY_DISPATCHER() &&
```

Comment:
maybe use asserts instead?

Reply:
If just change it to asserts, the GPDB won't even start (because, for instance, planner can be called with the utility role also).

As the final aim of the refactoring is to move ORCA to a shared lib, I'd like to keep other logic close to the original as much as possible.

Reply:
But `planner_hook = gp_orca_planner;` is set only in `_PG_init` when `IS_QUERY_DISPATCHER` and `Gp_role != GP_ROLE_DISPATCH`, i.e. same conditions.

Reply:
Gp_role can switch in the `InitPostgres()` [here](https://github.com/arenadata/gpdb/blob/98b00aef8bee397b5ee19380e18ee4951857133c/src/backend/utils/init/postinit.c#L1179) from `dispatch` to `utility`.
Thus the hook will be registered for an utility backend. And it worked this way before my change as well, as the ORCA was inited at the very beginning of `InitPostgres()`, before this role switch.

So we need the runtime check in the hook.

Reply:
> the hook will be registered for an utility backend

How?! The hook is now registered only when loading a shared library, and even then under additional conditions!

Reply:
We only need ORCA on the coordinator for planning.